shared/src/ledger/mod.rs

```diff
@@ -14,6 +14,521 @@ pub mod queries;
+            fee_amount: InputAmount::Unvalidated(token::DenominatedAmount {
+                amount: token::Amount::default(),
+                denom: NATIVE_MAX_DECIMAL_PLACES.into(),
+            }),
+            fee_payer: None,
+            fee_token: self.native_token().await,
+            nut: false,
+            code_path: PathBuf::from(TX_BRIDGE_POOL_WASM),
+            tx: self.tx_builder().await,
+        }
+    }
+
+    /// Make a ResignSteward builder from the given minimum set of arguments
+    async fn new_resign_steward(
+        &self,
+        steward: Address,
+    ) -> args::ResignSteward {
+        args::ResignSteward {
+            steward,
+            tx: self.tx_builder().await,
+            tx_code_path: PathBuf::from(TX_RESIGN_STEWARD),
+        }
+    }
+
+    /// Make a UpdateStewardCommission builder from the given minimum set of
+    /// arguments
+    async fn new_update_steward_rewards(
+        &self,
+        steward: Address,
+        commission: Vec<u8>,
+    ) -> args::UpdateStewardCommission {
+        args::UpdateStewardCommission {
+            steward,
+            commission,
+            tx: self.tx_builder().await,
+            tx_code_path: PathBuf::from(TX_UPDATE_STEWARD_COMMISSION),
+        }
+    }
+
+    /// Make a TxCustom builder from the given minimum set of arguments
+    async fn new_custom(&self, owner: Address) -> args::TxCustom {
+        args::TxCustom {
+            owner,
+            tx: self.tx_builder().await,
+            code_path: None,
+            data_path: None,
+            serialized_tx: None,
+        }
+    }
+
+    /// Sign the given transaction using the given signing data
+    async fn sign(
+        &self,
+        tx: &mut Tx,
+        args: &args::Tx,
+        signing_data: SigningTxData,
+    ) -> crate::sdk::error::Result<()> {
+        signing::sign_tx(*self.wallet_mut().await, args, tx, signing_data)
+    }
+
+    /// Process the given transaction using the given flags
+    async fn submit(
+        &self,
+        tx: Tx,
+        args: &args::Tx,
+    ) -> crate::sdk::error::Result<ProcessTxResponse> {
+        tx::process_tx(self, args, tx).await
+    }
+}
+
+/// Provides convenience methods for common Namada interactions
+pub struct NamadaImpl<'a, C, U, V, I>
+where
+    C: crate::ledger::queries::Client + Sync,
+    U: WalletIo,
+    V: ShieldedUtils,
+    I: Io,
+{
+    /// Used to send and receive messages from the ledger
+    pub client: &'a C,
+    /// Stores the addresses and keys required for ledger interactions
+    pub wallet: RwLock<&'a mut Wallet<U>>,
+    /// Stores the current state of the shielded pool
+    pub shielded: RwLock<&'a mut ShieldedContext<V>>,
+    /// Captures the input/output streams used by this object
+    pub io: &'a I,
+    /// The default builder for a Tx
+    prototype: args::Tx,
+}
+
+/// The Namada token
+pub const NAM: &str = "atest1v4ehgw36x3prswzxggunzv6pxqmnvdj9xvcyzvpsggeyvs3cg9qnywf589qnwvfsg5erg3fkl09rg5";
```

Comment:
Hard coded addresses give me nightmares since I've been working on the new genesis flow :sweat_smile: . Is there any way this can be looked up instead?

Reply:
I mean eventually it will have to be looked up, it will just fall on me.

Reply:
Nightmares are not good. Formerly I used to look it up in the wallet - but the SDK is frequently used in contexts where the wallet is empty (see https://github.com/anoma/namada-sdk-starter/pull/1 and also the benchmarks). Would it be valid to have an RPC endpoint that returns the native token address?

Reply:
That works for me.